Cleaner exit when user credentials are incorrect

craddm opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

- Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
- If you're not yet ready to merge, please mark this pull request as a **draft** and add `'[WIP]'` to the title
- Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

<!--
List any other PRs that must be merged before this one
-->

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

When the user rejects the current credentials when asked to confirm they are correct, code would still continue to run, resulting in numerous additional, unnecessary and misleading error messages.

In addition, when the user rejects the Graph API credentials as incorrect, there is currently no clear way to *change* those credentials.
Graph API log-in is, unlike Az CLI log-in, not done by the user at the command line.

This PR simply exits when the user confirms that the credentials are incorrect, and prints only informative error messages. 
In addition, it advises the user how to remove the cached Graph API credentials so as to be able to login with other credentials. 

### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #2280 
Closes #2153 

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

Tested locally

craddm committed:
linting

craddm committed:
Directly exit after credentials not confirmed

craddm committed:
Catch more specific exceptions rather than all exceptions

craddm committed:
Tell user how to remove cached Graph API credential

craddm committed:
Merge remote-tracking branch 'upstream/develop' into fix-error-message

craddm committed:
remove newline from error message

craddm committed:
Catch another specific error type

craddm committed:
Fix tests to reflect change in exception type

craddm committed:
fix linting

craddm committed:
fix linting

JimMadge commented on data_safe_haven/external/api/credentials.py:
```diff
@@ -144,8 +145,7 @@ def get_credential(self) -> TokenCredential:
             )
-            msg = "Error getting account information from Azure CLI."
-            raise DataSafeHavenAzureError(msg) from exc
+            raise typer.Exit(code=1) from exc
```
I don't think we should raise this exception anywhere except from Typer `command`s.

This module shouldn't depend on Typer.

JimMadge reviewed (dismissed):
I don't think this is how we should solve this. We want to catch the problem and raise a suitable exception, which is handled by the command function.

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into fix-error-message

JimMadge committed:
Revert some changes

JimMadge committed:
Adjust formatting

JimMadge committed:
Move target name to class var

JimMadge committed:
Simplify traceback

JimMadge commented:
Example output

<img width="643" alt="Screenshot 2024-11-20 at 10 02 17" src="https://github.com/user-attachments/assets/2246c373-99ed-40f9-8d17-4903dd1097cb">

JimMadge committed:
Fix Azure credential test

JimMadge committed:
Add stdout check to test

JimMadge committed:
Move tests path outside of default args

This makes the script more flexible, you can add options without having
to specify the test dir.

JimMadge committed:
Fix no application test

JimMadge committed:
Run lint:fmt

JimMadge reviewed (approved):
@craddm Could you take a look at this and say what you think?

I can't request a review from you as you opened the PR 😆

craddm commented:
LGTM
